File: python/binary-search/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        left, right = 0, len(matrix) - 1
        while left <= right:
            middle = (left + right) // 2
            if target <= matrix[middle][-1] and target >= matrix[middle][0]:
                row = middle
                left, right = 0, len(matrix[0]) - 1
                while left <= right:
                    middle = (left + right) // 2
                    if target == matrix[row][middle]:
                        return True
                    else:
                        if target < matrix[row][middle]:
                            right = middle - 1
                        else:
                            left = middle + 1

                break
            else:
                if target < matrix[middle][0]:
                    right = middle - 1
                else:
                    left = middle + 1

        return False

File: c++/binary-search/search-a-2d-matrix.cc
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int left_row = 0;
        int right_row = matrix.size() - 1;
        while (left_row <= right_row) {
            int middle_row = (left_row + right_row) / 2;
            if (target < matrix[middle_row][0]) {
                right_row = middle_row - 1;
            } else if (target > matrix[middle_row][matrix[middle_row].size() - 1]) {
                left_row = middle_row + 1;
            } else {
                int left = 0;
                int right = matrix[middle_row].size() - 1;
                while (left <= right) {
                    int middle = (left + right) / 2;
                    if (matrix[middle_row][middle] == target) {
                        return true;
                    } else if (matrix[middle_row][middle] < target) {
                        left = middle + 1;
                    } else {
                        right = middle - 1;
                    }
                }
                break;
            }
        }
        return false;
    }
};
